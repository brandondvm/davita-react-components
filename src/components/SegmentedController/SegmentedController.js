import React, { useState } from 'react';
import PropTypes from 'prop-types';

const SegmentedController = ({ controllerOptions = [], register }) => {
  const [ options, setOptions] = useState(controllerOptions);  

  const optionClick = selectedOption => () => {
    setOptions(options.map(option => ({
      checked: option.id === selectedOption.id,
      ...option
    })));
  }

  return (
    <div className="segmented-controller-component">      
      {         
        options.map((option, index) => 
        <div key={index} className="segmented-controller-component_item">
          <input defaultChecked={option.checked} className={`segmented-controller-component_input ${option.checked ? 'checked' : ''}`} ref={register} role="button" type="radio" id={option.id} name={option.name} value={option.value} />
          <label role="button" onClick={optionClick(option)} className="segmented-controller-component_item" htmlFor={option.id}>{option.label}</label>
        </div>
          
        )
      }
    </div>
  );
};

SegmentedController.propTypes = {
  controllerOptions: PropTypes.array,
  register: PropTypes.func
};

export default SegmentedController;
