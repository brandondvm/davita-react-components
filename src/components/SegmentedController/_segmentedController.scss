.segmented-controller-component {
  align-items: center;
  background-color: $palewhitegray;
  border-radius: 4px;
  display: flex;
  height: 32px;
  justify-content: center;
  padding: 2px 3px;

  &_input {
    height: 100%;
    margin: 0;
    position: absolute;
    visibility: hidden;
    width: 100%;

    &:checked ~ .segmented-controller-component_item {
      background-color: $white;
      border-radius: 4px;
      font-family: $bold-font;
    }
  }

  &_item {
    align-items: center;
    background-color: $palewhitegray;
    color: $onyx;
    display: flex;
    flex-basis: 0;
    flex-grow: 1;
    font-family: $regular-font;
    font-size: 14px;
    height: 100%;
    justify-content: center;
    position: relative;

    &_active {
      background-color: $white;
      border-radius: 4px;
      font-family: $bold-font;
    }

    &:hover {
      cursor: pointer;
    }

    &:focus {
      outline:none;
  
      &:active {
        outline:none;
      }
    }
  }
}
