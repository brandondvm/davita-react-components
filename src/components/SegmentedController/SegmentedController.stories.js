import React from 'react';

import SegmentedController from './SegmentedController';

const segmentedControllerItems = [
  { id: "option1", label: "option 1", name: "segmentedControllerOption", value: 1, checked: true },
  { id: "option2", label: "option 2", name: "segmentedControllerOption", value: 2, checked: false },
  { id: "option3", label: "option 3", name: "segmentedControllerOption", value: 3, checked: false }
];

export default {
  title: 'Segmented Controller',
  component: SegmentedController,
};

const Template = (args) => <div className="storybook-wrapper"><SegmentedController {...args} /></div>;

export const Default = Template.bind({});
Default.args = {
  controllerOptions: segmentedControllerItems,
};
