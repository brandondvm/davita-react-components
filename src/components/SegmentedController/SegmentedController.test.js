import enzyme from 'enzyme';
import React from 'react';

import SegmentedController from './SegmentedController';

const mockProps = {
  controllerOptions: [
    { id: 'bloodPressureOption1', label: 'Sitting', name: 'bloodPressureState', value: 'Sitting', checked: true },
    { id: 'bloodPressureOption2', label: 'Standing', name: 'bloodPressureState', value: 'Standing', checked: false },
    { id: 'bloodPressureOption3', label: 'Lying', name: 'bloodPressureState', value: 'lying', checked: false }
  ],
  register: jest.fn(),
};

function setup(passedProps = {}) {
  const updatedProps = {
    ...mockProps,
    ...passedProps,
  };

  return enzyme.shallow(<SegmentedController { ...updatedProps } />);
}

describe('Segmented Controller Component', () => {
  test('Component renders correctly', () => {
    const wrapper = setup();
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  test('Component should render 3 items', () => {
    const wrapper = setup();
    const itemsElements = wrapper.find('.segmented-controller-component_input');
    expect(itemsElements.length).toEqual(3);
  });

  test('Component should render 2 items', () => {
    const wrapper = setup({
      controllerOptions: [
        { id: 'bloodPressureOption1', label: 'Sitting', name: 'bloodPressureState', value: 'Sitting', checked: true },
        { id: 'bloodPressureOption2', label: 'Standing', name: 'bloodPressureState', value: 'Standing', checked: false }
      ],
    });
    const itemsElements = wrapper.find('.segmented-controller-component_input');
    expect(itemsElements.length).toEqual(2);
  });

  test('Component should render first item as active', () => {
    const wrapper = setup();
    const itemElement = wrapper.find('.segmented-controller-component_input').first();
    expect(itemElement.hasClass('checked')).toEqual(true);
  });

  test('Component should render last item as active', () => {
    const wrapper = setup({ 
      controllerOptions: [
        { id: 'bloodPressureOption1', label: 'Sitting', name: 'bloodPressureState', value: 'Sitting', checked: false },
        { id: 'bloodPressureOption2', label: 'Standing', name: 'bloodPressureState', value: 'Standing', checked: true },
      ],
     });
    const itemElement = wrapper.find('.segmented-controller-component_input').last();
    expect(itemElement.hasClass('checked')).toEqual(true);
  });
});
