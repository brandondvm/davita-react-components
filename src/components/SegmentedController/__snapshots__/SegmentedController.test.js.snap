// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segmented Controller Component Component renders correctly 1`] = `
<div
  className="segmented-controller-component"
>
  <div
    className="segmented-controller-component_item"
  >
    <input
      className="segmented-controller-component_input checked"
      defaultChecked={true}
      id="bloodPressureOption1"
      name="bloodPressureState"
      role="button"
      type="radio"
      value="Sitting"
    />
    <label
      className="segmented-controller-component_item"
      htmlFor="bloodPressureOption1"
      onClick={[Function]}
      role="button"
    >
      Sitting
    </label>
  </div>
  <div
    className="segmented-controller-component_item"
  >
    <input
      className="segmented-controller-component_input "
      defaultChecked={false}
      id="bloodPressureOption2"
      name="bloodPressureState"
      role="button"
      type="radio"
      value="Standing"
    />
    <label
      className="segmented-controller-component_item"
      htmlFor="bloodPressureOption2"
      onClick={[Function]}
      role="button"
    >
      Standing
    </label>
  </div>
  <div
    className="segmented-controller-component_item"
  >
    <input
      className="segmented-controller-component_input "
      defaultChecked={false}
      id="bloodPressureOption3"
      name="bloodPressureState"
      role="button"
      type="radio"
      value="lying"
    />
    <label
      className="segmented-controller-component_item"
      htmlFor="bloodPressureOption3"
      onClick={[Function]}
      role="button"
    >
      Lying
    </label>
  </div>
</div>
`;
